URI Paths Canonicalization

pmd1nh opened the pull request:
https://github.com/openliberty/open-liberty/issues/20414

pmd1nh committed:
URI Paths Canonicalization

pmd1nh commented:
#build

#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.session_fat,com.ibm.ws.session_fat_lWAS_servlet.3.1.toleration,com.ibm.ws.session.cache_fat,com.ibm.ws.session.cache_fat_config,com.ibm.ws.session.cache_fat_config_infinispan,com.ibm.ws.session.cache_fat_infinispan,com.ibm.ws.session.cache_fat_infinispan_container,com.ibm.ws.session.cache_provider_fat

pmd1nh commented:
#build

pmd1nh committed:
FAT for URI paths canonicalization

pmd1nh commented:
#build

 #spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.session_fat,com.ibm.ws.session_fat_lWAS_servlet.3.1.toleration,com.ibm.ws.session.cache_fat,com.ibm.ws.session.cache_fat_config,com.ibm.ws.session.cache_fat_config_infinispan,com.ibm.ws.session.cache_fat_infinispan,com.ibm.ws.session.cache_fat_infinispan_container,com.ibm.ws.session.cache_provider_fat

jantley-ibm commented on dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops:
```diff
@@ -1614,3 +1614,8 @@ duplicate.samesite.attribute.for.cookie.name.useraction=Check the deployment des
 duplicate.reported.for.samesite.attribute.cookie.name.explanation=A cookie name cannot be configured for multiple SameSite attribute values. This cookie name has already been identified as a duplicate.
 duplicate.reported.for.samesite.attribute.cookie.name.useraction=Check the deployment descriptor configuration to make sure the cookie name is configured for only one of the following SameSite attribute values: 'SameSiteCookies_Lax', 'SameSiteCookies_None', or 'SameSiteCookies_Strict'. 
+
+bad.request.uri\:.{0}=CWWWC0005I: The request URI has invalid or bad encoded characters: {0}
```
change
or bad encoded characters:
to
or improperly encoded characters:

jantley-ibm commented on dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops:
```diff
@@ -1614,3 +1614,8 @@ duplicate.samesite.attribute.for.cookie.name.useraction=Check the deployment des
 duplicate.reported.for.samesite.attribute.cookie.name.useraction=Check the deployment descriptor configuration to make sure the cookie name is configured for only one of the following SameSite attribute values: 'SameSiteCookies_Lax', 'SameSiteCookies_None', or 'SameSiteCookies_Strict'. 
+
+bad.request.uri\:.{0}=CWWWC0005I: The request URI has invalid or bad encoded characters: {0}
+bad.request.uri\:.{0}.explanation=The request URI has some combination of invalid path parameter or dot segment or bad encoded characters.
```
Change to:
The request URI has invalid path parameters, invalid dot segments, or improperly encoded characters.

jantley-ibm commented on dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops:
```diff
@@ -1614,3 +1614,8 @@ duplicate.samesite.attribute.for.cookie.name.useraction=Check the deployment des
+
+bad.request.uri\:.{0}=CWWWC0005I: The request URI has invalid or bad encoded characters: {0}
+bad.request.uri\:.{0}.explanation=The request URI has some combination of invalid path parameter or dot segment or bad encoded characters.
+bad.request.uri\:.{0}.useraction=Verify the request UIR does not contain any invalid dot and path parameter combination or bad encoded characters.
```
Change to:
Verify that the request URI does not contain an invalid dot or path parameter combination or any improperly encoded characters.

jantley-ibm reviewed (commented):
ID review comments attached. When we agree on changes, I will add the ID Reviewed label. Let me know if you have any questions or concerns. Thanks -

pmd1nh commented:
full security test as I think OpenID may use fragment

#build
#spawn.fullfat.buckets=com.ibm.websphere.security_fat,com.ibm.ws.collective.security_fat,com.ibm.ws.collective.security_zfat,com.ibm.ws.ejbcontainer.security.3.2_fat,com.ibm.ws.ejbcontainer.security_fat,com.ibm.ws.ejbcontainer.security_fat.features,com.ibm.ws.security.audit_fat.config,com.ibm.ws.security.audit_fat.example,com.ibm.ws.security.audit_fat.features,com.ibm.ws.security.auth.data_fat,com.ibm.ws.security.authentication.builtin_fat,com.ibm.ws.security.authentication.tai_fat,com.ibm.ws.security.authorization.saf_zfat,com.ibm.ws.security.client_fat_java7,com.ibm.ws.security.context_fat,com.ibm.ws.security.credentials.saf_zfat,com.ibm.ws.security.csiv2_fat,com.ibm.ws.security.java2sec_fat,com.ibm.ws.security.jwt_fat.builder,com.ibm.ws.security.jwt_fat.consumer,com.ibm.ws.security.oauth_fat,com.ibm.ws.security.oidc.client_fat,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.quickstart_fat,com.ibm.ws.security.registry.saf_zfat,com.ibm.ws.security.registry_fat,com.ibm.ws.security.saml.sso_fat.config,com.ibm.ws.security.saml.sso_fat.config.mapToUserRegistry,com.ibm.ws.security.saml.sso_fat.endpoint.samlmetadata,com.ibm.ws.security.saml.sso_fat.jaxrs,com.ibm.ws.security.saml.sso_fat.jaxrs.config,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP,com.ibm.ws.security.spnego_fat,com.ibm.ws.security.thread.identity_fat,com.ibm.ws.security.thread.zos.context_zfat,com.ibm.ws.security.thread.zos_zfat,com.ibm.ws.security.token.ltpa_fat,com.ibm.ws.security.utility_fat,com.ibm.ws.security.wim.scim_fat,com.ibm.ws.security.wim.scim_zfat,com.ibm.ws.security_fat,com.ibm.ws.security_fat.features,com.ibm.ws.webcontainer.security.feature_fat,com.ibm.ws.webcontainer.security.jacc.1.5_fat,com.ibm.ws.webcontainer.security_fat,com.ibm.ws.webcontainer.security_fat.features,com.ibm.ws.webcontainer.security_fat.passwordutil,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat.spnego,com.ibm.ws.security.spnego_fat,com.ibm.ws.security.spnego_fat.1,com.ibm.ws.security.spnego_fat.2

pmd1nh commented:
@jantley-ibm I updated with the suggestion.  Thanks!

pmd1nh commented:
all security passed. 4 are invalid



Invalid test bucket - com.ibm.ws.security.audit_fat.config | 1 | 1 | 0 |   |  
-- | -- | -- | -- | -- | --
Invalid test bucket - com.ibm.ws.security.oidc.client_fat | 1 | 1 | 0 |   |  
Invalid test bucket - com.ibm.ws.security.oidc.client_fat.config | 1 | 1 | 0 |   |  
Invalid test bucket - com.ibm.ws.security.social_fat.LibertyOP

pmd1nh commented:
SkipForRepeat EE10 JMX tests that have URI not compliant with Servlet 6.0

#build

pmd1nh commented:
Confirmed with Ut Le that the above security are full fat spec; so look like security full fat are fine.

pmd1nh commented:
Issue to re-enable Rest tests
https://github.com/OpenLiberty/open-liberty/issues/23271

posmith commented:
L2 message review completed

pmd1nh commented:
Design issue:
https://github.com/OpenLiberty/open-liberty/issues/23279

pmd1nh commented:
Updated to reject control character

#build

#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat

pmd1nh commented:
SERVLET5_URI_PATH_COMPATIBILITY, true by default testing

#build

pmd1nh commented:
#build

pmd1nh commented:
Updated error report construction

#build

pmd1nh commented:
Updated - 2 properties:      

SKIP_ENCODED_CHAR_VERIFICATION (default is false)
COLLAPSE_EMPTY_SEGMENTS_IN_URI (default is false)

#build

pmd1nh commented:
full security

#build
#spawn.fullfat.buckets=com.ibm.websphere.security_fat,com.ibm.ws.collective.security_fat,com.ibm.ws.collective.security_zfat,com.ibm.ws.ejbcontainer.security.3.2_fat,com.ibm.ws.ejbcontainer.security_fat,com.ibm.ws.ejbcontainer.security_fat.features,com.ibm.ws.security.audit_fat.config,com.ibm.ws.security.audit_fat.example,com.ibm.ws.security.audit_fat.features,com.ibm.ws.security.auth.data_fat,com.ibm.ws.security.authentication.builtin_fat,com.ibm.ws.security.authentication.tai_fat,com.ibm.ws.security.authorization.saf_zfat,com.ibm.ws.security.client_fat_java7,com.ibm.ws.security.context_fat,com.ibm.ws.security.credentials.saf_zfat,com.ibm.ws.security.csiv2_fat,com.ibm.ws.security.java2sec_fat,com.ibm.ws.security.jwt_fat.builder,com.ibm.ws.security.jwt_fat.consumer,com.ibm.ws.security.oauth_fat,com.ibm.ws.security.oidc.client_fat,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.quickstart_fat,com.ibm.ws.security.registry.saf_zfat,com.ibm.ws.security.registry_fat,com.ibm.ws.security.saml.sso_fat.config,com.ibm.ws.security.saml.sso_fat.config.mapToUserRegistry,com.ibm.ws.security.saml.sso_fat.endpoint.samlmetadata,com.ibm.ws.security.saml.sso_fat.jaxrs,com.ibm.ws.security.saml.sso_fat.jaxrs.config,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP,com.ibm.ws.security.spnego_fat,com.ibm.ws.security.thread.identity_fat,com.ibm.ws.security.thread.zos.context_zfat,com.ibm.ws.security.thread.zos_zfat,com.ibm.ws.security.token.ltpa_fat,com.ibm.ws.security.utility_fat,com.ibm.ws.security.wim.scim_fat,com.ibm.ws.security.wim.scim_zfat,com.ibm.ws.security_fat,com.ibm.ws.security_fat.features,com.ibm.ws.webcontainer.security.feature_fat,com.ibm.ws.webcontainer.security.jacc.1.5_fat,com.ibm.ws.webcontainer.security_fat,com.ibm.ws.webcontainer.security_fat.features,com.ibm.ws.webcontainer.security_fat.passwordutil,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat.spnego,com.ibm.ws.security.spnego_fat,com.ibm.ws.security.spnego_fat.1,com.ibm.ws.security.spnego_fat.2

pmd1nh commented:
COLLAPSE_EMPTY_SEGMENTS_IN_URI (default is true) i.e alignment with the spec.

#build
#spawn.fullfat.buckets=com.ibm.websphere.security_fat,com.ibm.ws.collective.security_fat,com.ibm.ws.collective.security_zfat,com.ibm.ws.ejbcontainer.security.3.2_fat,com.ibm.ws.ejbcontainer.security_fat,com.ibm.ws.ejbcontainer.security_fat.features,com.ibm.ws.security.audit_fat.config,com.ibm.ws.security.audit_fat.example,com.ibm.ws.security.audit_fat.features,com.ibm.ws.security.auth.data_fat,com.ibm.ws.security.authentication.builtin_fat,com.ibm.ws.security.authentication.tai_fat,com.ibm.ws.security.authorization.saf_zfat,com.ibm.ws.security.client_fat_java7,com.ibm.ws.security.context_fat,com.ibm.ws.security.credentials.saf_zfat,com.ibm.ws.security.csiv2_fat,com.ibm.ws.security.java2sec_fat,com.ibm.ws.security.jwt_fat.builder,com.ibm.ws.security.jwt_fat.consumer,com.ibm.ws.security.oauth_fat,com.ibm.ws.security.oidc.client_fat,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.quickstart_fat,com.ibm.ws.security.registry.saf_zfat,com.ibm.ws.security.registry_fat,com.ibm.ws.security.saml.sso_fat.config,com.ibm.ws.security.saml.sso_fat.config.mapToUserRegistry,com.ibm.ws.security.saml.sso_fat.endpoint.samlmetadata,com.ibm.ws.security.saml.sso_fat.jaxrs,com.ibm.ws.security.saml.sso_fat.jaxrs.config,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP,com.ibm.ws.security.spnego_fat,com.ibm.ws.security.thread.identity_fat,com.ibm.ws.security.thread.zos.context_zfat,com.ibm.ws.security.thread.zos_zfat,com.ibm.ws.security.token.ltpa_fat,com.ibm.ws.security.utility_fat,com.ibm.ws.security.wim.scim_fat,com.ibm.ws.security.wim.scim_zfat,com.ibm.ws.security_fat,com.ibm.ws.security_fat.features,com.ibm.ws.webcontainer.security.feature_fat,com.ibm.ws.webcontainer.security.jacc.1.5_fat,com.ibm.ws.webcontainer.security_fat,com.ibm.ws.webcontainer.security_fat.features,com.ibm.ws.webcontainer.security_fat.passwordutil,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat.spnego,com.ibm.ws.security.spnego_fat,com.ibm.ws.security.spnego_fat.1,com.ibm.ws.security.spnego_fat.2

pmd1nh commented:
Based; updated rest handler web.xml to add context-param for SKIP_ENCODED_CHAR_VERIFICATION 

#build

pmd1nh commented:
removed unused method that moved to filterManager

#build

pmd1nh commented:
minor change to reorg the trace(). no need for PB with this.

pmd1nh commented:
CL change to ignore.xml some errors after the correct versions has been added to bnd.bnd and package-info.

https://github.ibm.com/websphere/WS-CD-Open/pull/28372

pmd1nh commented:
@jantley-ibm I added a new property to the metatype.  Can you  please review its info?  Thanks!

dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties

jantley-ibm commented on dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -185,4 +185,7 @@ setHtmlContentTypeOnError.name=Disable text/html content type on error responses
\ No newline at end of file
+excludeAllHandledTypesClasses.desc=When this property is set to true, during startup, ServletContainerInitializer implementors that use the HandlesTypes annotation do not receive classes that are specified as HandlesTypes parameters.
+
+skipEncodedCharVerification.name=Skip verify the encoded char %23, %2E, %2F, and %5C in request URI
```
suggested rewrite:
skipEncodedCharVerification.name=Skip verification of the %23, %2E, %2F, and %5C encoded characters in the request URI

jantley-ibm commented on dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -185,4 +185,7 @@ setHtmlContentTypeOnError.name=Disable text/html content type on error responses
+excludeAllHandledTypesClasses.desc=When this property is set to true, during startup, ServletContainerInitializer implementors that use the HandlesTypes annotation do not receive classes that are specified as HandlesTypes parameters.
+
+skipEncodedCharVerification.name=Skip verify the encoded char %23, %2E, %2F, and %5C in request URI
+skipEncodedCharVerification.desc=A request is rejected if any of the encoded char %23, %2E, %2F, or %5C is found in the request URI. When this property is set to true, the URI is not checked for these encoded chars.
```
suggested rewrite:
skipEncodedCharVerification.desc=A request is rejected if the request URI contains the %23, %2E, %2F, or %5C encoded characters. When this property is set to true, the request URI is not checked for these encoded characters.

jantley-ibm reviewed (commented):
ID review comments attached for recent addition. Let me know if you have any questions or concerns about the suggested wording.

pmd1nh commented:
@jantley-ibm updated with suggested texts.  Thanks!

jantley-ibm commented:
> @jantley-ibm updated with suggested texts. Thanks!

Thanks, looks good. ID reviewed tag already applied.

pmd1nh commented:
metatype name can't be more than 50 characters.

#build

pnicolucci commented on dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -185,4 +185,7 @@ setHtmlContentTypeOnError.name=Disable text/html content type on error responses
+excludeAllHandledTypesClasses.desc=When this property is set to true, during startup, ServletContainerInitializer implementors that use the HandlesTypes annotation do not receive classes that are specified as HandlesTypes parameters.
+
+skipEncodedCharVerification.name=Skip verification of encoded characters in URI
+skipEncodedCharVerification.desc=A request is rejected if the request URI contains the %23, %2E, %2F, or %5C encoded characters. When this property is set to true, the request URI is not checked for these encoded characters.
```
Phu this is specific to Servlet 6.0 right? I think we need to be clear here for what Servlet feature version this metatype applies to.

pmd1nh commented:
updated javadoc.zip version

#build

pmd1nh commented:
#build

pmd1nh commented:
rebased
#build

pmd1nh commented:
Updated javadoc.zip in 

dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.servlet-servletSpi2.0.feature

#build

pmd1nh commented:
rebased and updated to EPL-2.0

#build

#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.session_fat,com.ibm.ws.session_fat_lWAS_servlet.3.1.toleration,com.ibm.ws.session.cache_fat,com.ibm.ws.session.cache_fat_config,com.ibm.ws.session.cache_fat_config_infinispan,com.ibm.ws.session.cache_fat_infinispan,com.ibm.ws.session.cache_fat_infinispan_container,com.ibm.ws.session.cache_provider_fat

pmd1nh commented:
rebased to pickup and jmxconnector change

#build

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/WebContainer.java:
```diff
@@ -211,6 +210,9 @@ public Object run() {
     // Servlet 4.0 : Must be static since referenced from static method
     protected static CacheServletWrapperFactory cacheServletWrapperFactory;
+    
+    //Servlet 6.0 - do not load it early
```
What does this comment mean? Just that you're not initializing the variable here?

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/WebContainer.java:
```diff
@@ -808,18 +812,43 @@ public void handleRequest(IRequest req, IResponse res, VirtualHost vhost, Reques
+                            + "), (encoding=" + encoding + ")");
             }
-            // End 293696 ServletRequest.getPathInfo() fails WASCC.web.webcontainer
+            // End 293696 ServletRequest.getPathInfo() fails WASCC.web.webcontainereve
```
What is `webcontainereve`?

pmd1nh replied:
remove the eve

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/WebContainer.java:
```diff
@@ -1677,6 +1706,207 @@ public static synchronized Iterator<ServletContainerInitializer> getServletConta
+    
+    //Servlet 6.0
+    //https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0.html#uri-path-canonicalization
+    //This method will be after the URI is decoded
```
This method will be **called** after...

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/filter/WebAppFilterManager.java:
```diff
@@ -1142,10 +1144,44 @@ else if (isTraceOn && logger.isLoggable(Level.FINE)) {
+                                alreadyVerifiedEncodedChar = true;        //either way, skip subsequent check in filterDefined
+                            }
+                            catch (IOException ioe) {
+                                logger.logp(Level.INFO, CLASS_NAME, "invokeFilters", "servletWrapper. Bad request - sending 400 [" + ioe.getMessage() + "]");
```
This should be translated if we're putting out a INFO message in the logs

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/filter/WebAppFilterManager.java:
```diff
@@ -1142,10 +1144,44 @@ else if (isTraceOn && logger.isLoggable(Level.FINE)) {
+                                alreadyVerifiedEncodedChar = true;        //to skip check in filterDefined later on
+                            }
+                            catch (IOException ioe) {
+                                logger.logp(Level.INFO, CLASS_NAME, "invokeFilters", "genericServletWrapper. Bad request - sending 400 [" + ioe.getMessage() + "]");
```
This should be translated if we're putting out an INFO message in the logs

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/filter/WebAppFilterManager.java:
```diff
@@ -1293,6 +1344,24 @@ else if (servletWrapper instanceof GenericServletWrapper){
+                                            alreadyVerifiedEncodedChar = true;  //nothing after this, but just incase
+                                        }
+                                        catch (IOException ioe) {
+                                            logger.logp(Level.INFO, CLASS_NAME, "invokeFilters", "DefaultExtensionProcessor. Bad request - sending 400 [" + ioe.getMessage() + "]");
```
This should be translated if we're putting out an INFO message in the logs

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/webapp/WebAppConfiguration.java:
```diff
@@ -2169,4 +2169,38 @@ public void setModuleResponseEncoding(String encoding) {
+        }
+    }
+
+    public boolean skipVerifyEncodedCharInURI() {
```
Since we're returning the value of this variable a better method name would be `isSkipVerifyEncodedCharInURI`

pmd1nh replied:
change to isSkipVerifyEncodedCharInURI

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/webapp/WebAppConfiguration.java:
```diff
@@ -2169,4 +2169,38 @@ public void setModuleResponseEncoding(String encoding) {
+     * 
+     * true - skip checking for %23 , %2e , %2f , %5c in URI
+     */
+    public void setSkipEncodedCharVerification() {
```
We only care about this context-parameter and metaType in Servlet 6.0 correct? Where are we verifying that it isn't used when not using the servlet-6.0 feature?

pmd1nh replied:
this setSkipEncodedCharVerification() is only called from the new 
WebAppConfiguratorHelper60() which is active only for servlet6.0+.

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/WebContainer.java:
```diff
@@ -808,18 +812,43 @@ public void handleRequest(IRequest req, IResponse res, VirtualHost vhost, Reques
+                    path = canonicalizeURI(decodedReqURI);
+                }
+                catch (IOException ioe) {
+                    logger.logp(Level.INFO, CLASS_NAME, "handleRequest", "canonicalize sending 400 [" + ioe.getMessage() + "]");
```
If we're logging an INFO message to the logs it should be translated.

pmd1nh replied:
change to  Level.FINE

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/filter/WebAppFilterManager.java:
```diff
@@ -1202,6 +1238,21 @@ else if (servletWrapper instanceof GenericServletWrapper){
+                        alreadyVerifiedEncodedChar = true;      // skip checking in the DefaultExtension
+                    }
+                    catch (IOException ioe) {
+                        logger.logp(Level.INFO, CLASS_NAME, "invokeFilters", "filtersDefined. Bad request - sending 400 [" + ioe.getMessage() + "]");
```
If we're logging an INFO message to the logs this should be translated.

pnicolucci commented on dev/com.ibm.ws.webcontainer/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -345,6 +342,12 @@
             description="%excludeAllHandledTypesClasses.desc"
             required="false" type="Boolean" />
+            
+        <!--com.ibm.ws.webcontainer.-->
```
This needs to be beta guarded. We should open an issue to remove the beta guard when ready.

pmd1nh commented:
#build

pmd1nh commented:
Address feature checker by correct the version  of com.ibm.wsspi.webcontainer in the /dev/wlp-jakartaee-transform/rules/jakarta-versions*.properties files

#build

pnicolucci commented on dev/io.openliberty.webcontainer.servlet.6.0.internal_fat/test-applications/URIPathCanonicalizationTest.war/resources/jsp/errorPage.jsp:
```diff
@@ -0,0 +1,23 @@
+<!--
+  Copyright (c) 2023 IBM Corporation and others.
```
Can you fix this formatting?

pmd1nh commented:
only need to update the Export-package.

#build

pnicolucci commented on dev/io.openliberty.webcontainer.servlet.6.0.internal/src/io/openliberty/webcontainer60/osgi/container/config/WebAppConfiguratorHelper60.java:
```diff
@@ -1,14 +1,11 @@
 /*******************************************************************************
- * Copyright (c) 2022 IBM Corporation and others.
+ * Copyright (c) 2023 IBM Corporation and others.
```
Why was the date changed and a range not used ?

pmd1nh replied:
updated

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/webapp/WebAppErrorReport.java:
```diff
@@ -178,6 +176,11 @@ else if (rootCause instanceof UnavailableException) {
         }
+        else if (rootCause instanceof IOException
+                        && th.getMessage() != null
+                        && th.getMessage().contains("CWWWC0005I")) {     //Servlet 6.0 added
```
No way this can happen unless with Servlet 6 correct? No need to add a check?

pmd1nh replied:
No, IOException has been thrown before.  This check ensure that a 400 is returned only if the IOException coming from the new path.  Without this check, a 400 won't be set.  or without the last check "th.getMessage().contains("CWWWC0005I", any IOException will get a 400 which may not be correct.

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/filter/WebAppFilterManager.java:
```diff
@@ -1573,4 +1642,52 @@ private boolean findAndSetServletInfo(IFilterMapping fmInfo) {
+                    logger.logp(Level.FINE, CLASS_NAME, METHOD_NAME, "Bad Request : " + message);
+
+                throw new IOException(nls.getFormattedMessage("bad.request.uri:.{0}", new Object[] { ((uri.length() > 128) ? (uri.substring(0, 127)) : uri) }, 
+                                "Bad request URI") + " . " + message);
```
Shouldn't the `message` part of this be translated as well?

pmd1nh replied:
add messages for translation

pnicolucci commented on dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -187,4 +184,7 @@ setHtmlContentTypeOnError.name=Disable text/html content type on error responses
+excludeAllHandledTypesClasses.desc=When this property is set to true, during startup, ServletContainerInitializer implementors that use the HandlesTypes annotation do not receive classes that are specified as HandlesTypes parameters.
+
+skipEncodedCharVerification.name=Skip verification of encoded characters in URI
+skipEncodedCharVerification.desc=A request is rejected if the request URI contains the %23, %2E, %2F, or %5C encoded characters. When this property is set to true, the request URI is not checked for these encoded characters. This property is available starting in servlet 6.0, and default value is false.
```
add a new line at end of file

pmd1nh commented:
#build

pmd1nh commented:
@jantley-ibm We are finalizing this change before merging.  Can you please re-review the messages as they have been updated?  Thanks!

jantley-ibm commented on dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops:
```diff
@@ -1616,3 +1613,12 @@ duplicate.samesite.attribute.for.cookie.name.useraction=Check the deployment des
+
+uri.has.fragment.character=URI has a fragment [#] character; encoded [%23] or not.
+uri.has.dot.character=URI has encoded dot [%2E] character.
+uri.has.forwarslash.character=URI has encoded forwardslash [%2F] character.
```
change so that forward slash is two words:
uri.has.forwarslash.character=URI has encoded forward slash [%2F] character.

IBM Style guide has "backslash" as one word and "forward slash" as two.

pmd1nh replied:
updated

jantley-ibm reviewed (commented):
ID has one small comment: forward slash is two words.

pnicolucci commented:
#libby

pmd1nh commented:
#build

pmd1nh commented:
com.ibm.ws.ejbcontainer.injection_fat bucket has one suite failed: LookupOverrideTest



com.ibm.ws.ejbcontainer.injection_fat FAT tests | 433 | 28 | 2 |   |  
-- | -- | -- | -- | -- | --
com.ibm.ws.ejbcontainer.injection.fat.tests.InjectionMiscTest | 11 | 0 | 0 | 3 s | 1
com.ibm.ws.ejbcontainer.injection.fat.tests.InjectionMIXTest | 255 | 0 | 0 | 40 s | 2
com.ibm.ws.ejbcontainer.injection.fat.tests.InjectionXMLTest | 51 | 0 | 0 | 8 s | 3
com.ibm.ws.ejbcontainer.injection.fat.tests.LookupOverrideTest | 29 | 28 | 1 | 1 s | 4
com.ibm.ws.ejbcontainer.injection.fat.tests.RemoteInjectionTest | 86 | 0 | 0 | 17 s | 6
junit.framework.TestSuite | 1 | 0 | 1 | 0 s | 5
 
  
Rerun the child build with only this bucket, all pass:

com.ibm.ws.ejbcontainer.injection_fat FAT tests | 432 | 0 | 0 |   |  
-- | -- | -- | -- | -- | --
com.ibm.ws.ejbcontainer.injection.fat.tests.InjectionMiscTest | 11 | 0 | 0 | 3 s | 1
com.ibm.ws.ejbcontainer.injection.fat.tests.InjectionMIXTest | 255 | 0 | 0 | 38 s | 2
com.ibm.ws.ejbcontainer.injection.fat.tests.InjectionXMLTest | 51 | 0 | 0 | 7 s | 3
com.ibm.ws.ejbcontainer.injection.fat.tests.LookupOverrideTest | 29 | 0 | 0 | 6 s | 4
com.ibm.ws.ejbcontainer.injection.fat.tests.RemoteInjectionTest | 86 | 0 | 0 | 13 s | 5
 

pmd1nh commented:
Requested message translation.
